### DISCUSSION TITLE
Redis memory leak with Statamic nocache regions?

### DISCUSSION THREAD

**Discussion by JorisOrangeStudio** [statamic/cms#10634] [Troubleshooting] [answered]
We're experiencing a Redis memory leak on our Statamic websites (8GB server). Memory usage increases by ~0.5-1GB daily, eventually maxing out. Clearing the application cache of the websites temporarily resolves the issue.

- Using Static half page caching via Redis
- Redis is the main memory consumer (in the graph below, redis was using 4-5GB from the 8GB on the 14th of August) 
- Numerous `nocache::region.*` keys present (300-600 KB each, no TTL set) and amount of keys keeps increasing

<img width="928" alt="image" src="https://github.com/user-attachments/assets/f754971f-44ba-4057-89b2-7d2a7f4751a1">

Are we missing out on some invalidation for these nocache regions? Or what could be the reason of the increasing memory 

Any insights or solutions would be greatly appreciated.

**Comment by ryanmitchell** [comment]
You should use the select="xx" param on the nocache tags if you aren't already: <https://statamic.dev/tags/nocache#performance>.
This will ensure the region only includes the data you actually need and prevents these sorts of build ups.

**Reply by JorisOrangeStudio** [reply]
So in this example, adding `select="@auto"` to the `nocache` tag will improve this behaviour? 
```html
{{ nocache }}
    {{ collection:offers query_scope="offers_query" }}
        <div class="swiper-slide">
          {{ partial:components/picture :image="images:0" }}
          <div>{{ heading }}</div>
          <div>{{ content }}</div>
          <a href="{{ permalink }}">{{ trans:strings.more_info }}</a>
        </div>
    {{ /collection:offers }}
{{ /nocache }}
```

The nocache tag will still create separate keys in Redis for each instance then right? If so, is there a way to set TTLs for these nocache regions to prevent indefinite growth? Or is that not recommended to do?

**Reply by ryanmitchell** [reply]
You can try auto - personally I prefer to be explicit. So I this case I would have `{{ nocache select="id" }}` so it selects the entry id.

**Reply by JorisOrangeStudio** [reply]
Thanks Ryan! I'll give it a try. Didn't knew that `{{ nocache select="id" }}` is already enough for this loop. So if I'd be explicit, 
`{{ nocache select="images|heading|content|permalink" }}` is not needed here? 

> The nocache tag will still create separate keys in Redis for each instance then right? If so, is there a way to set TTLs for these nocache regions to prevent indefinite growth? Or is that not recommended to do?

What about the TTL of those nocache regions, do you know more about this behaviour?

**Reply by ryanmitchell** [reply]
No they arent needed as they come from your collection loop, the select is only what data you want to be made available to your nocache region (think of it like a partial).

Theres no TTL as such, they exist until the static cache is cleared.

**Reply by martyf** [reply]
Can you explain that again Ryan? I don’t understand why “id” is all that’s needed in this example.

**Reply by ryanmitchell** [reply]
`images|heading|content|permalink` are all provided by the collection loop so they dont need to be supplied to the nocache region.

`id` isnt really needed in the output, its just something thats not going to change per visit/session so its a good candidate for using in the select to minimise region build up.

**Reply by martyf** [reply]
And that’s the ID of the entry that is being loaded, right?

**Reply by ryanmitchell** [reply]
Yep. You could use url too I guess

**Reply by martyf** [reply]
There you go, learn something new every day - thanks for that Ryan.

**Reply by JorisOrangeStudio** [reply]
Thanks Ryan, so `id` is just there to scope down the region, since the current entry ID is not used by the collection loop either?

One more question regarding this:
> Theres no TTL as such, they exist until the static cache is cleared.


I've enabled cache expiry (6 hours) on the half page caching like this: 
```
        'half' => [
            'driver' => 'application',
            'expiry' => 360,
        ],
``` 
For the `static-cache:responses:*` keys there is actually a TTL for 6 hours added, but the `nocache::region.*` keys don't have this. That means that only on manual `static:clear` it will clear the indefinite growth regions right?

**Reply by ryanmitchell** [reply]
Not sure on that, probably one for @jasonvarga to answer

**Reply by JorisOrangeStudio** [reply]
I've looked into the NoCache source, not sure if this is the correct place, but it indeed looks like the regions are set forever: 

https://github.com/statamic/cms/blob/dfa5e1da1ae5def03db93cc8581f14011a9592ca/src/StaticCaching/NoCache/Session.php#L153

Is this done with a reason🤔 I can image that after invalidating within a certain time and requesting the region again will result in a `RegionNotFound` exception? I'm not sure about the 'lookup/get' behaviour of regions.

**Comment by jasonvarga** [comment]
Do you use `nocache` on pages that aren't statically cached?

**Reply by JorisOrangeStudio** [reply]
No, we don't have any exclusions set up in the static cache config

**Reply by JorisOrangeStudio** [reply]
I was wondering if there is any update on this. Over the past 5 days I noticed that the same occurred (Now it is around ±6GB in nocache regions on Redis). Let me know if I need to create an issue for it :) 
<img width="947" alt="image" src="https://github.com/user-attachments/assets/31c14496-c5bd-4c0e-adbf-b115e3efabc0">

**Reply by jasonvarga** [reply, marked as answer]
Keep an eye on https://github.com/statamic/cms/pull/10671
